accepted/future-releases/nnbd/feature-specification.md

```diff
@@ -1236,18 +1248,92 @@ the `LEGACY_ERASURE` of the types.
+literals which would compare equal via the definition of runtime type equality
+given above will be canonicalized to the same type object.  The choice of what
+type object to canonicalize to is arbitrary in that placement of legacy
+modifiers in type literals is not otherwise observable in the language.  Any
+consistent choice of placement of legacy modifiers is hence an equally valid
+implementation.
+
+Note that the choice of canonicalization for type literals does not depend
+directly on whether sound or unsound null safety semantics are in use.  However,
+soundly null safe programs will never contain legacy types, and hence the
+erasure of legacy type modifiers is never necessary when running in sound mode.
+
+#### Constant instances
+
+In both sound and unsound null checking, canonicalization of generic object
+instances is performed relative to the normal forms of their generic type
+arguments.  That is, an instance of `C<T0>` canonicalizes to the same object as
+`C<T1>` if `T0` and `T1` have the same normal form (up to the identity of bound
+variables), and the objects are otherwise identical.
 
 With sound null checking, all generic const constructors and generic const
-literals are evaluated using the actual type arguments provided, whether legacy
-or non-legacy.  This ensures that with sound null checking, the final
-consistent semantics are obeyed.
+literals are evaluated using the actual type arguments provided.  This ensures
+that with sound null checking, the final consistent semantics are obeyed.
+
+With unsound null checking, all generic const constructors and generic const
+literals are additionally treated as if all type arguments passed to them were
+legacy types regardless of whether the constructed class is defined in a legacy
+library or not, and regardless of whether the constructor invocation or literal
+occurs in a legacy library or not.  That is, a constructor invocation or object
+literal with generic type parameters `Ti` is treated as if the parameters were
+**LEGACY_TYPE**('Ti'). Implementations may choose to eagerly normalize the type
+arguments before applying the legacy rewrite, as desired.
+
+This ensures that const objects which appear identical in the syntax continue to
+canonicalize consistently across legacy and opted-in libraries.  Unlike the case
+of type literals, the choice to rewrite the type arguments is not arbitrary,
+since the outcome of constant instance checks and casts may depend on the legacy
+status of the type arguments.
+
+The Dart static analysis tool does not distinguish between sound and unsound
+checking mode, and hence it is expected that there will be some small level of
+infidelity in the constant evaluation semantics in the analyzer.  In practice,
+programs that observe object identity at compile time are almost non-existent,
+and so it may prove most pragmatic in the analyzer to always use the provided
+type arguments.
```

Comment:
I can see the problem.
```dart
// null safe code.
class C<T> {
  const C(Object o) : assert(o is T Function(T));
}
int? f(int? x) => x;
const breaks = C<int>(f); // Throws in sound mode, not in unsound mode.
```
I can't see a good solution. I don't like the analyzer doing having a decidedly wrong semantics (one which matches neither unsound mode nor sound mode), but since we are erring on the side of detecting problems in sound mode programs, it's probably the best available alternative in the long run. And statically checking identity of constants in potentially constant code is likely to be rare.


Maybe the analyzer could constant-evaluate null safe code twice, one in sound mode and one in unsound mode. It'll still look weird to have "potential, but not certain, compile-time error".

Reply:
This kind of analyzer infidelity in mixed mode has been baked in for a year plus, I'm not too worried about it.  In practice, I think users will be happy if the const evaluation is performed according to the library semantics: that is, in an opted in library, the above would throw, but if the `const breaks` definition were in an opted out library, it would not throw (since the type would be `int*`).  I actually think that from a behavior standpoint, this is superior to the CFE semantics.  Unfortunately, it is inconsistent with the separate desire to have two instances of `<int>[]` compare as identical, even if one appears in an opted in library and one does not.  So I'm proposing to give the CFE the consistent identity semantics, but permit the analyzer to be looser with identity (which is basically never observed at compile time in practice) but relatively strict with evaluation semantics.  Though actually, after discussions with @scheglov and @bwilkerson today, I believe that the analyzer can implement accurate identity semantics.